### PR TITLE
Vectorized Interval.__and__, Interval.__or__, and Interval.difference

### PR DESCRIPTION
**Pull request by milosobral**
### What has changed:
- Before: interval ops used Python-heavy traversal patterns with repeated appends.
- Now: interval ops use numpy-friendly logic (searchsorted, array slicing, and vectorized merges).
- Why this helps: large interval sets run much faster, especially for __and__ and __or__, where the old approach spent more time in Python loops. This makes a very large difference for real life slice calls.

### Why slice benchmarks do not show huge gains directly:
The direct slice benchmarks are useful, but they do not fully match production behavior. I think it's because of the following:

In the synthetic benchmark, a lot of time is spent in array work and HDF5 overhead. That masks part of the Python-level overhead reductions from the optimizations. So the benchmark number for slice() looks modest even when the real pipeline gets faster. It also seems to depend quite a bit on the case, we could maybe improve the benchmarking suite to cover more realitic cases.

### Real-world impact
In real usage (large Data objects with many split fields), I've see:

- about 10x faster slice.
- fetching a single batch is about 3 times faster
- about 2.5x speedup when generating 10 batches of size 64

### Benchmark results: 

Benchmark | 041f06c90f (µs) | Working Tree (µs) | Speedup
-- | -- | -- | --
Data.slice() (lazy, realistic) | 21752.432 | 18622.578 | 1.17x
Data.slice() (in-memory) | 604.553 | 384.890 | 1.57x
IrregularTimeSeries.slice() | 55.508 | 50.041 | 1.11x
Interval.slice() | 5.840 | 5.699 | 1.02x
Interval.__and__ (1k&single) | 550.042 | 25.773 | 21.34x
Interval.__and__ (1k&100) | 737.445 | 600.303 | 1.23x
Interval.__or__ (1k\|100) | 4341.434 | 44.595 | 97.35x
Interval.difference (1k-100) | 3992.252 | 3821.029 | 1.04x
ArrayDict.keys() x100k | 0.653 | 0.661 | 0.99x
LazyInterval access (10 attrs) | 1139.397 | 1117.498 | 1.02x

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved interval operations (difference, intersection, union) to handle empty, identical, containment, adjacent, and point cases more reliably.

* **Refactor**
  * Reworked internal implementations for faster, array-oriented processing and added early-exit paths for simple cases to improve performance and robustness.

* **Tests**
  * Added extensive tests covering edge cases, multi-segment interactions, point intervals, and error conditions.

* **Documentation**
  * Added CHANGELOG entry summarizing the interval operation improvements.

### CODE REVIEW

File: tests/test_interval.py [hunk 1]
```diff
@@ -579,6 +579,96 @@ def easy_symmetric_check(I1, I2, Iexp, op):
...
+    # identical intervals
+    I1 = Interval.from_list([(1.0, 3.0), (5.0, 7.0)])
+    Iexp = Interval.from_list([(1.0, 3.0), (5.0, 7.0)])
+    easy_symmetric_check(I1, I1, Iexp, op)
+
+    # multiple containments in a row
+    I1 = Interval.from_list([(0.0, 20.0)])
+    I2 = Interval.from_list([(1.0, 3.0), (5.0, 7.0), (9.0, 11.0)])
+    Iexp = Interval.from_list([(0.0, 20.0)])
+    easy_symmetric_check(I1, I2, Iexp, op)
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply tests/test_interval.py]: Although, I am not sure if we want to keep this behavior..
Do you think this behavior of merging adjacent intervals in `__or__` is reasonable?
It is incorrect from a math point-of-view
